xdsl/transforms/stencil_bufferize.py

```diff
@@ -0,0 +1,435 @@
+        if not op.res:
+            return
+
+        bounds = cast(TempType[Attribute], op.res[0].type).bounds
+
+        dests = [
+            AllocOp(result_types=[field_from_temp(cast(TempType[Attribute], r.type))])
+            for r in op.res
+        ]
+        operands = [
+            (
+                BufferOp.create(
+                    operands=[o],
+                    result_types=[field_from_temp(o.type)],
+                )
+                if isa(o.type, TempType[Attribute])
+                else o
+            )
+            for o in op.operands
+        ]
+
+        loads = [
+            LoadOp(operands=[d], result_types=[r.type]) for r, d in zip(op.res, dests)
+        ]
+
+        new = ApplyOp(
+            operands=[operands, dests],
+            regions=[Region(Block(arg_types=[SSAValue.get(a).type for a in operands]))],
+            result_types=[[]],
+            properties={"bounds": bounds},
+        )
+        rewriter.inline_block(
+            op.region.block,
+            InsertPoint.at_start(new.region.block),
+            new.region.block.args,
+        )
+
+        rewriter.replace_matched_op(
+            [*(o for o in operands if isinstance(o, Operation)), *dests, new, *loads],
+            [SSAValue.get(l) for l in loads],
+        )
+
+
+def walk_from(a: Operation) -> Generator[Operation, Any, None]:
+    """
+    Walk through all operations recursively inside a or its block.
+    """
+    while True:
+        yield from a.walk()
+        if a.next_op is None:
+            break
+        a = a.next_op
+
+
+def walk_from_to(a: Operation, b: Operation):
+    """
+    Walk through all operations recursively inside a or its block, until b is met, if
+    ever.
+    """
+    for o in walk_from(a):
+        if o == b:
+            return
+        yield o
+
+
+class LoadBufferFoldPattern(RewritePattern):
+    """
+    The underlying buffer of values resulting from a load of a buffer is the buffer
+    loaded from. That is, if nobody annoyed the said buffer in-between.
+
+    If someone did, we should probably insert a copy there. Right now, we just bail out.
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: BufferOp, rewriter: PatternRewriter):
+
+        handle = op.res
+        # If this is a value-semantic buffer, we can't fold it
+        if not isinstance(handle.type, FieldType):
+            return
+
+        temp = op.temp
+        load = temp.owner
+        # We are interested in folding buffers of loaded values
+        if not isinstance(load, LoadOp):
+            return
+
+        underlying = load.field
+
+        # TODO: propery analysis of effects in between
+        # For illustration, only fold a single use of the handle
+        # (Requires more boilerplate to analyse the whole live range otherwise)
+        uses = handle.uses.copy()
+        if len(uses) > 1:
+            return
+        user = uses.pop().operation
```

Comment:
it's probably not safe to call `pop()` if there is a control-flow path in which you return without rewriting the op.

Reply:
Hence the `.copy()` above :wink: 

There must be more poetic ways, propositions welcome!